docs/source/traits_user_manual/internals.rst

```diff
@@ -177,6 +177,122 @@ Attribute deletion
 Attribute deletion (``del obj.name``) goes through the same code path as
 attribute set operations. Most ``CTrait`` types do not permit deletion.
 
+Traits Containers
+-----------------
+
+Traits has the ability to watch for changes in standard Python containers:
+lists, dictionaries and sets.  To achieve this Traits provides special
+subclasses of the standard Python classes that can validate elements and can
+fire trait notifications when the contents change.  These classes are,
+respectively, |TraitList|, |TraitDict| and |TraitSet| (not to be confused with
+the deprecated Trait Handlers of the same names).
+
+In addition to being able to take an appropriate value to initialize the
+container (such as a sequence or mapping), these container subclasses also
+take keyword-only arguments for validators (either a single item validator for
+|TraitList| and |TraitSet|, or key and value validators for |TraitDict|) and
+notifiers.
+
+These classes were introduced in Traits 6.1 and the implementation details
+described here are provisional and may change.
+
+Backwards Compatibility
+~~~~~~~~~~~~~~~~~~~~~~~
+
+In practice, most traits containers are instances of |TraitListObject|,
+|TraitDictObject| and |TraitSetObject| which are subclasses that are
+created by the validators of |List|, |Dict| and |Set| traits respectively and
+supplied with appropriate element validators that wrap the standard trait
+validators so that constructs like ``List(Str)`` can work.
+
+These objects are designed to be API compatible with the classes of the
+same name from Traits 6.0 and before, and so have different constructors and
+may behave slightly differently from the base classes in some cases.
+
+In particular |TraitListObject| classes can have restrictions on their
+length, which is not part of the base |TraitList| API.
+
+All of these backward compatibility classes are strongly tied to a particular
+object and trait, and are not designed to operate as stand-alone entities.
+These relationships are required to support the `object.*_items`-style event
+traits, but complicate the implementation, for example by having to use
+weak references to the object, and having to take this additional structure
+into account when serializing and deserializing.
+
+It is a long-term goal to phase out the use of these backwards compatibility
+classes.
+
+Container Validators
+~~~~~~~~~~~~~~~~~~~~
+
+Container validators are callables that are expected to take a single value
+and either return a validated value or raise a TraitError.  The default
+validators simply pass the input value directly through, but validators can
+potentially do much more.   Validators are expected to be idempotent:
+``validate(validate(x)) == validate(x)`` should hold as long as
+``validate(x)`` does not raise an error.
+
+For example, the following validator casts the list items to integers,
```

Comment:
Nitpick, but can we spell this as "converts" instead of "casts" (and similarly elsewhere). C has casts. Python doesn't. (Yes, I know this is a losing battle.)